[Ready]Sticky ban in game interface

MrsTonedOne opened the pull request:
This allows admins to add and remove dream deamon sticky bans.

Minor credit to /vg/ as an example for the primary SetConfig() code

This marks the stickybans it adds so that it can only be used to edit/remove sticky bans it created.
- [x] Viewing sticky bans
- [x] Adding sticky bans
- [x] Editing sticky bans
- [x] Removing sticky bans
- [x] Delinking ckeys from sticky bans
- [x] Edit isbanned() to better handle deamon bans
- [ ] Exempting ckeys from matching a given sticky ban
- [x] Making admins immuned to stickybans but only if placed by the code
- [x] Clean up code
- [ ] Make the interface less shit.
- [x] ~~Add a prompt for making a perma ban a sticky ban when banning (how /vg/ does it)~~ (decided against this for now)

Close enough.

Other notes:

isbanned() cleaned up, it logs sticky ban matches, and better handles admins being exempt from bans
Adminbans now still work once the admin is demoted.
Admins bypassing a ban because they are an admin is logged and announced to all admins including the one who walked past it.
Admins are now exempt from host bans. (this only applies to host bans for ss13, global host bans (where the 'apply to this game only' checkbox is not checked (defaults to not checked)) do not trigger isbanned() and thus, admins can not bypass them, no matter what we do.)

Added a system to queue a message for a client, to be shown next time they connect, this was needed because isbanned() is called before the client is created, so if you want to send a message to an admin, letting them know they just walked pass a matching ban, you have to do it this way.

MrsTonedOne committed:
Sticky ban in game interface

Core0verload commented:
How BYOND's internal ban system works?

SconesC commented:
Fucking awesome

MrStonedOne commented:
>How BYOND's internal ban system works?

Stupidly robust. Sticky bans (dream deamon bans with the stickyban checkbox checked) have stopped most ban evaders who use vms or the cid randomizer dll.

The most notable example is terbs. We added a sticky ban on his original account that had been banned for over a year, after he came in to irc bragging about how he was gonna ban evade now that he moved and got a new computer, isp, and everything...

Think about that, new isp, new computer, new location, no trace right?

Sticky ban caught him trying evade under the username purity ring.

Its fucking black magic.

Miauw62 commented:
all the turds in the byond codebase probably compressed it into a black hole of coding, and sticky bans happened to be at the center.

Core0verload commented:
I suspect some sort of combined hardware/software/provider fingerprinting, including embedded IE parameters fingerprinting, because I tried to use embedded IE fingerprinting at least once and it mostly worked. Will test it on local server anyway. Hope these bans are not *too* sticky.

freerealestate commented:
oh i thought this was an interface to ban sticky

maybe next time

Jalleo commented:
He probably used that account on his old computer but elsewhere in byond.
Also there needs to be sanity checks with this mso funny thimg happened a couple o weeks ago due to a badly setup sticky ban.

MrStonedOne commented:
>Also there needs to be sanity checks with this mso funny thimg happened a couple o weeks ago due to a badly setup sticky ban.

I know about that, and already accounted for them

>maybe he just had more than one computer when he got the new one and connected with the alternative one thinking he hadn't connected to the server with it in the past?

<SplitQuote>

>He probably used that account on his old computer but elsewhere in byond.

Terbs knows what he's doing, i doubt he'd fuck up like that.

Chnkr commented:
> Its fucking black magic.
Not really, i just tested how sticky bans actually work on localhost and i could evade them pretty easily without any effort, it just requires a extra step when registering a new account.

Jalleo commented:
Welp our secret weapon aint so poweful anymore then.

optimumtact commented:
hello

Razharas commented:
@MrStonedOne

MrStonedOne commented:
I need to get back on this, I kept getting stuck at reworking the logic of isbanned() so I think I'll scrap a few features for now to make it manageable.

Razharas commented:
Weekly ping of activity reminder
@MrStonedOne

Niknakflak commented:
Hoping this gets finished soon @MrStonedOne

tedward1337 commented:
@mrstonedone tick tock

lollerderby commented:
Great work @MrStonedOne I can't wait to see this implemented @MrStonedOne.

LanCartwright commented:
I want this.

Niknakflak commented:
@mrstonedone

Niknakflak commented:
@mrstonedone

Jalleo commented:
Just give it time to mature and be ready treat it like a whiskey.

Razharas commented:
I want to still be young when this is ready to merge so
@MrStonedOne

Shadowlight213 commented:
@Mrstonedone

MrsTonedOne committed:
Cleans up isbanned() and stickyban handling
isbanned() cleaned up, it logs sticky ban matches, and better handles admins being exempt from bans
Adminbans now still work once the admin is demoted.
Admins bypassing a ban because they are an admin is logged and announced to all admins including the one who walked past it.
Admins are now exempt from host bans. (this only applies to host bans for ss13, global host bans (where the 'apply to this game only' checkbox is not checked (defaults to not checked)) do not trigger isbanned() and thus, admins can not bypass them, no matter what we do.)

Added a system to queue a message for a client, to be shown next time they connect, this was needed because isbanned() is called before the client is created, so if you want to send a message to an admin, letting them know they just walked pass a matching ban, you have to do it this way.

MrsTonedOne committed:
MERGE CONFLICTS, MY ONLY WEAKNESS
How did you know?

duncathan commented:
@MrStonedOne

Markolie commented on code/modules/admin/IsBanned.dm:
```diff
@@ -63,10 +28,13 @@
+			if (admin)
+				log_admin("The admin [key] has been allowed to bypass a matching ban on [.["key"]]")
+				message_admins("<span class='adminnotice'>The admin [key] has been allowed to bypass a matching ban on [.["key"]]</span>")
+				addclientmessage(ckey,"<span class='adminnotice'>You have been allowed to bypass a matching ban on [.["key"]]")
```
Missing </span> here.

Markolie commented on code/modules/admin/IsBanned.dm:
```diff
@@ -99,12 +67,46 @@
+				else
+					log_admin("The admin [key] has been allowed to bypass a matching ban on [pckey]")
+					message_admins("<span class='adminnotice'>The admin [key] has been allowed to bypass a matching ban on [pckey]</span>")
+					addclientmessage(ckey,"<span class='adminnotice'>You have been allowed to bypass a matching ban on [pckey]")
```
Closing span

Markolie commented on code/modules/admin/IsBanned.dm:
```diff
@@ -99,12 +67,46 @@
+		if (admin)
+			log_admin("The admin [key] has been allowed to bypass a matching host/sticky ban")
+			message_admins("<span class='adminnotice'>The admin [key] has been allowed to bypass a matching host/sticky ban</span>")
+			addclientmessage(ckey,"<span class='adminnotice'>You have been allowed to bypass a matching host/sticky ban")
```
Closing span

Razharas commented:
@MrStonedOne 
So is this ready or are you going to achieve some more goals from op post?

MrsTonedOne committed:
Adds missing closing span tags to admin ban bypass notices.

MrStonedOne commented:
@Razharas This is as ready as it will ever be. The interface is clunky, but I suck at simple html interfaces, so it will just have to do.

Razharas commented:
Ill pretend i reviewed it so it can be tested on live servs for actual problem finding

d3athrow commented:
once you know how a sticky ban works it's trivial to bypass, but most people will never figure out how.
